Limit theft/destroy secondary objectives to items which at least one person cares about

Jacquerel opened the pull request:
## About The Pull Request

This PR adds a new field `item_owner` to `datum/objective_item` and populates it for most objective items.
The function of this field is that the objective will fail to generate if there isn't at least one player in the game with one of those roles. This limits theft objectives to targets which are likely to be carried around, missed if they become absent, or at least vaguely have someone standing close to.
This has only been applied toprogression traitor secondary objectives. 

While doing this I also culled a number of deprecated or tedious objectives which we didn't use any more, such as every "steal a stack of x" objective. And the plasma tank one. And the slime core one. 
I also deleted any which nobody would miss, chiefly "an empty AI card". There's plenty of those things and if one vanished nobody would notice let alone be inconvenienced.
If anyone ever wants to bring them back for soul reasons it will be retained in the commit history, that's what git is for.

The list is actually not particularly long now, so if people have good ideas for more things to add to expand it again... well they could have done that any time in the past 10 years but they can still do it now as long as it isn't to fill it back up with things nobody will miss if they are gone.
For myself, I added the robotics lab mech crowbar to the list as it fulfils most of the criteria of "fire axe" but with the benefit of there only being one on most maps.

I also removed the `difficulty` var which was undocumented, not set for most objectives (all "low risk" objectives had a default difficulty of `9001`), and unused.

Also I think this is my 100th PR so you have to be nice to me.

## Why It's Good For The Game

Traitor objectives (and antagonist objectives in general) should:
- Present you with a somewhat engaging task
- Encourage you to interact with at least one other player (at least by "attempting to sneak past them")
- Move the state of the round forward in some way, usually by creating inconvenience for at least one other person

Many current progression traitor objectives do not meet all (and some perhaps not even any) of the above criteria and I am planning on making a series of changes in order to address this. Expect to see this paragraph copied/pasted into some other PRs probably.

This one is the simplest: Stealing an item from someone who literally isn't in the game does not antagonise anyone. So we won't ask anyone to do it.

## Changelog

:cl:
balance: Traitor secondary objectives to steal or destroy an item are now associated with a particular "owner" or "victim" and won't generate themselves if nobody would actually be victimised by the crime.
add: Traitors can sometimes be tasked with stealing and destroying the roboticist's big crowbar, if there is one
/:cl:

Jacquerel committed:
Limit theft objectives to items which at least one person cares about

Sealed101 commented:
> I also removed the difficulty var which was undocumented, not set for most objectives (all "low risk" objectives had a default difficulty of 9001), and unused.

wtf soul removal

coiax commented on code/game/gamemodes/objective_items.dm:
```diff
@@ -2,23 +2,45 @@
+	if (!length(item_owner))
+		return TRUE
+	for (var/mob/player as anything in GLOB.player_list)
+		if (!(player.mind?.assigned_role.title in item_owner))
+			continue
+		if (is_centcom_level(player.z))
+			continue
```
Can you reprase this in a manner where the logic is easier to parse? I'd say combine this logic into a single "if this job exists, and they're not on Centcom; then true" branch, rather than the negative gauntlet.

coiax commented on code/game/gamemodes/objective_items.dm:
```diff
@@ -66,6 +71,7 @@
 	name = "a fire axe"
 	targetitem = /obj/item/fireaxe
 	excludefromjob = list(JOB_CHIEF_ENGINEER,JOB_STATION_ENGINEER,JOB_ATMOSPHERIC_TECHNICIAN)
+	item_owner = list(JOB_CHIEF_ENGINEER,JOB_STATION_ENGINEER,JOB_ATMOSPHERIC_TECHNICIAN)
```
Does the fire axe also include the one on the Bridge (on most maps?). Could command staff care about the fire axe?

Jacquerel replied:
Yes, I'm not all too sure what to do about that one to be honest because those are two very different locations so no list of owners is going to reliably cover both.
Maybe that's fine in this case on account of it being a bulky and difficult-to-conceal weapon.

Jacquerel replied:
That did give me an idea though

Jacquerel committed:
Simplifies loop
Actually removes these items from destroy objectives
Expand list of people who can't steal a fire axe
Adds big crowbar objective

coiax commented on code/game/gamemodes/objective_items.dm:
```diff
@@ -34,47 +54,56 @@
+	item_owner = list(JOB_ROBOTICIST)
+	exists_on_map = TRUE
+
+/obj/item/fireaxe/add_stealing_item_objective()
```
I think you mean `/obj/item/crowbar/mechremoval`, right?

Jacquerel committed:
Oopsie

Mothblocks commented on code/modules/antagonists/traitor/objectives/steal.dm:
```diff
@@ -161,6 +146,9 @@ GLOBAL_DATUM_INIT(steal_item_handler, /datum/objective_item_handler, new())
 		if(!target.TargetExists())
 			qdel(target)
 			continue
+		if(!target.owner_exists())
+			qdel(target)
+			continue
 		if(role.title in target.excludefromjob)
 			qdel(target)
 			continue
```
Can we clean this up so that this all can become something like:

```dm
if (!target_is_valid(target))
    qdel(target)
    continue
```

KittyNoodle commented:
dislike how this might interact with lowpop, your pretty much only gonna get contractor objectives, which are a pain without the baton

Mothblocks commented:
Get a stun baton then idk

Jacquerel commented:
> dislike how this might interact with lowpop, your pretty much only gonna get contractor objectives, which are a pain without the baton

It's your job as a traitor to make the round interesting. Farming reputation by breaking into empty rooms on lowpop isn't interesting. 
If this prompts you to do something else, that's an intentional feature.

Jacquerel committed:
Consolidates shared procs into one proc on the antagonist datum

Higgin commented:
Great change, thank you.

Watermelon914 commented on code/game/gamemodes/objective_items.dm:
```diff
@@ -2,26 +2,55 @@
+/datum/objective_item/proc/valid_objective_for(list/potential_thieves)
+	if(!target_exists() || !owner_exists())
+		return FALSE
+	for (var/datum/mind/possible_thief in potential_thieves)
```
```suggestion
	for (var/datum/mind/possible_thief as anything in potential_thieves)
```

Watermelon914 commented:
You should make it so that traitors can't get the objective if the person of interest is also dead.

Jacquerel committed:
Theft item owners don't count if they are dead

Jacquerel committed:
Skip mind type check

Watermelon914 commented on code/game/gamemodes/objective.dm:
```diff
@@ -580,16 +580,14 @@ GLOBAL_LIST_EMPTY(possible_items)
-				if(M.current.mind.assigned_role.title in possible_item.excludefromjob)
-					continue check_items
-			approved_targets += possible_item
+	for(var/datum/objective_item/possible_item in GLOB.possible_items)
+		if(!possible_item.valid_objective_for(owners))
+			continue
+		if(possible_item.objective_type != OBJECTIVE_ITEM_TYPE_NORMAL)
+			continue
+		if(!is_unique_objective(possible_item.targetitem,dupe_search_range))
+			continue
+		approved_targets += possible_item
```
do note that I think this could be seriously problematic on lowpop for non-traitors since their steal objectives are permanent. I'm also introducing a change that gives traitors back their normal steal objectives so this could possibly lead to steal objectives actually being free objectives, which I don't think should be the case.

Watermelon914 replied:
I'd rather that permanent objectives have steal objectives for jobs that aren't flagged up right now rather than them having a free objective. That's what I mean.
Target-specific filters should only apply to temporary objectives that traitors are given.

Jacquerel replied:
I understand the reasoning but I would _prefer_ to apply it to all antagonists anyway if possible.
My reasoning is that there are still several objectives in the pool which _don't_ have an owner requirement, specifically:

- Vault Documents
- The Nuclear Bomb Core
- Supermatter Sliver
- The Blackbox
- Functioning AI (this handily turns itself into a murder objective if there is no AI)

These are all pretty fun "heist" style things in my opinion, are unlikely to be invalidated when joining late, and will be valid to appear on any round.

Would it be acceptable as an alternative that I just remove the Captain/HoS presence requirement from some of their items so that they're always available and leave it such that it still doesn't generate CMO/RD/CE theft objectives if there isn't one?

The Captain's Laser especially is probably still fine with no Captain as its location makes it a fun heist and then after stealing it the temptation to use it is very strong.
The HoS laser too is behind the entire security wing and while that's often completely empty on lowpop it gives a sufficient obstacle to get in and similarly then leaves you with the temptation to out yourself by carrying it around.

Jacquerel replied:
We discussed this on Discord and I'll just leave it only for secondary objectives.

Jacquerel committed:
Only apply restrictions to progression traitor objectives
